clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -967,6 +998,62 @@ private Record nextFetchedRecord() {
             }
             return records;
         }
+
+        private boolean isBatchAborted(RecordBatch batch, Set<Long> abortedPids, Queue<FetchResponse.AbortedTransaction> abortedTransactions) {
+           /* When in READ_COMMITTED mode, we need to do the following for each incoming entry:
+            *   0. Check whether the pid is in the 'abortedPids' set && the entry does not include an abort marker.
+            *      If so, skip the entry.
+            *   1. If the pid is in aborted pids and the entry contains an abort marker, remove the pid from
+            *      aborted pids and skip the entry.
+            *   2. Check lowest offset entry in the abort index. If the PID of the current entry matches the
+            *      pid of the abort index entry, and the incoming offset is no smaller than the abort index offset,
+            *      this means that the entry has been aborted. Add the pid to the aborted pids set, and remove
+            *      the entry from the abort index.
+            */
+            if (isolationLevel == IsolationLevel.READ_COMMITTED) {
+                FetchResponse.AbortedTransaction nextAbortedTransaction = abortedTransactions.peek();
+                if (abortedPids.contains(batch.producerId())
+                        || (nextAbortedTransaction != null && nextAbortedTransaction.producerId == batch.producerId() && nextAbortedTransaction.firstOffset <= batch.baseOffset())) {
```

Comment:
Not sure if that helps in this case, but the general idea would be to push some logic outside of Fetcher and into classes that are easier to unit test.